library/std/src/ffi/os_str.rs

```diff
@@ -667,6 +667,44 @@ impl OsStr {
         s.as_ref()
     }
 
+    /// Converts a slice of bytes to an OS string slice without checking that the string contains
+    /// valid `OsStr`-encoded data.
+    ///
+    /// See the [module's toplevel documentation about conversions][conversions] for safe,
+    /// cross-platform [conversions] from/to native representations.
+    ///
+    /// # Safety
+    ///
+    /// `OsStr`'s encoding is an unspecified superset of UTF-8 and callers must
+    /// pass in bytes that originated as a mixture of validated UTF-8 and bytes from
+    /// [`OsStr::as_os_str_bytes`].  The bytes from `OsStr::as_os_str_bytes` may be split either
+    /// immediately before or immediately after some valid non-empty UTF-8 substring
```

Comment:
There is a really critical piece left out here: that this constructor is only safe to call with the result of `OsStr::as_os_str_bytes` from within the same "process space." In other words, callers *must not* rely on the bytes produced by `as_os_str_bytes` to always result in constructing a valid `Osstr` via `from_os_str_bytes_unchecked`. This is because the specific encoding used is an implementation detail and may change.

It's also because WTF-8 should not be used as an interchange format. See http://simonsapin.github.io/wtf-8/#the-wtf-8-encoding and http://simonsapin.github.io/wtf-8/#intended-audience.

So for example, adding Serde impls that use this pair of methods to make serializing `OsString` possible would be a huge mistake and potentially lock us into never making changes to the encoding.

Reply:
It looks like you do mention this somewhat in `as_os_str_bytes`, but I think it needs to be explicitly part of the safety contract for the constructor.

Reply:
> It looks like you do mention this somewhat in as_os_str_bytes, but I think it needs to be explicitly part of the safety contract for the constructor.

Good call on having this in the `from`.

>  is only safe to call with the result of OsStr::as_os_str_bytes from within the same "process space."
> ...
> So for example, adding Serde impls that use this pair of methods to make serializing OsString possible would be a huge mistake and potentially lock us into never making changes to the encoding.

To double check, my wording is around "same rust version built for the same platform" which is describing the underling inputs that can change things rather than listing more general do's and don'ts and dealing with the edge cases.  Is there a preference on which wording approach I take?

Reply:
Yeah hmmmm. I think I like you're wording better, although it is perhaps a bit looser than mine.

My guess is that we have some wording for this particular type of concept elsewhere, but I'm not sure where.

I feel like my preference would be the most conservative thing possible. "within the same process" is the idea I'd like to go with to start personally. Otherwise, with your wording, I think one could justifying serializing these bytes so long as they don't change their compiler version. I'm not sure I'd want to encourage that.

Reply:
So adding a "for example, sending this over the network or storing this in a file will likely violate these rules"?  I can see the use for that.

Reply:
Yes, I like phrasing like that. Perfect. :)

Reply:
Pushed a commit that tries to integrate this.  Not fully happy with the wording and can see this having further tweaks but can't think of it atm.